frontend/src/app/visualization/ngram/ngram.component.ts

```diff
@@ -196,51 +200,46 @@ export class NgramComponent extends ParamDirective implements OnChanges {
     }
 
     loadGraph() {
-        this.isLoading.emit(true);
-
         this.lastParameters = _.clone(this.currentParameters);
         const cachedResult = this.getCachedResult(this.currentParameters);
 
         if (cachedResult) {
             this.onDataLoaded(cachedResult);
         } else {
-            this.visualizationService
-                .getNgramTasks(
-                    this.queryModel,
-                    this.corpus,
-                    this.visualizedField.name,
-                    this.currentParameters
-                )
-                .then((response) => {
-                    this.tasksToCancel = response.task_ids;
-                    return this.apiService.pollTasks<NgramResults>(
-                        response.task_ids
-                    );
-                })
-                .then(([result]) => {
-                    this.tasksToCancel = undefined;
-                    this.cacheResult(result, this.currentParameters);
-                    this.onDataLoaded(result);
-                })
-                .catch(this.onFailure.bind(this));
+            this.visualizationService.getNgramTasks(
+                this.queryModel, this.corpus, this.visualizedField.name,
+                this.currentParameters).then(
+                    response => {
+                        this.tasksToCancel = response.task_ids;
+                        // tasksToCancel contains ids of the parent task and its subtasks
+                        // we are only interested in the outcome of the parent task (first in array)
+                        const poller$ = this.apiService.pollTasks([this.tasksToCancel[0]], this.stopPolling$);
+                        poller$.subscribe({
+                            error: (error) => this.onFailure(error),
+                            next: (result: SuccessfulTask<NgramResults[]>) => this.onDataLoaded((result).results[0]),
+                            complete: () => this.apiService.abortTasks({ task_ids: this.tasksToCancel })
```

Comment:
It looks odd to cancel tasks that have seemingly just completed.

Reply:
This line indeed means we fire off a request to abort a completed task. Slightly inefficient, but it does avoid having to call `abortTasks` everywhere in tandem with stopPolling. Perhaps the complete handler could still check the type of result though to avoid this.